### ISSUE TITLE
getting zeros in validation during training

### ISSUE THREAD

**Issue by nikky4D** [realtime-intelligence/evrt-detr#6]
Can anyone help me debug my training session? I'm seeing zeros in my validation and am not sure why. 

I am attempting a train session on my own collected data: 130 samples in train, 7 samples in validation, 7 samples in test, with 6 classes. I'm using synthetic data generated from color videos following Prophesee website frames to events simulator here: [From Frames to Events](https://docs.prophesee.ai/stable/tutorials/ml/dataset_creation/video_to_event_simulator.html)

I've verified that the bounding boxes in the original color input are correct in the event stream .dat files before conversion. To do this, I'm using prophesee-automotive-dataset-toolbox to plot a single .dat file and the accompanying npy file, delta_t = 5000microseconds. I can see that many frames in the video have multiple bounding boxes, which will be cleaned up in the preprocessing psee_to_frames.py

However, when doing training from scratch for the first step, the output only shows 0 across the validation metrics. I only change the location of the dataset, and the number of classes. Here is a sample after more than 40 epochs

<img width="1234" height="271" alt="Image" src="https://github.com/user-attachments/assets/0295b780-4da0-4508-8b1a-2dd5ea3846d0" />

And after 80 epochs, here is my validation history: 

<img width="580" height="229" alt="Image" src="https://github.com/user-attachments/assets/86bf9723-0c8e-4ec8-9fcd-57fa34fc5fe3" />

the equivalent training history csv 

<img width="629" height="931" alt="Image" src="https://github.com/user-attachments/assets/d2f62036-efa3-4b9a-944d-cd206b72ee78" />

Here is my training config for train_gen4_rtdetr_presnet50.py up to epoch 58
[setup_for_debugging.txt](https://github.com/user-attachments/files/21128282/setup_for_debugging.txt)

**Comment by usert5432** [comment]
Hi @nikky4D,

There may be a few things that could cause the behavior. For instance, from the training log it seems like there are 65 training steps per a single epoch, while in the original training there are 1000 training iterations per epoch. So, it would take 1000 / 65 ~ 15 times more epochs to achieve the same performance as with the original training.

Another potential issue is that the "130 samples in train" may be too small dataset to achieve good performance. If possible, it may be beneficial to start with a pre-trained model on a gen4 dataset and fine-tune on your dataset. I can elaborate on this option if you prefer.

Another issue may be with the evaluator. We relied on the official prophesee evaluation code for the gen4 dataset. First, it applies the following filters before running the bounding box inference:
```
    'gen4' : {                                                                                                          
        'min_box_diag' : 60,                                                    
        'min_box_side' : 20,                                                    
        'skip_from_start_us' : 5e5,                                             
    },                                                                          
```
If your bounding boxes are smaller than 60 in diagonal or have a side less than 20 they will be filtered out.

There may be a few more things to investigate, but please let me know how would you prefer to proceed.

**Comment by nikky4D** [comment]
Thanks for the response. I appreciate the help. 

Could you elaborate on the finetuning and evaluation process? what is the command to start with a pretrained model for finetuning? Any advice on how best to finetune?

For evaluation, what files would I need to modify to change or turn off those values for evaluation? In the preprocessing in psee_to_frames.py, I only use the crop-to-fov mangling, and no other type of mangling. Also, I view the images of the event reps produced and the bounding boxes survive the psee_to_frames.py process. Here is the config file, train_gen4_rtdetr_presnet50.py, I use this for eval: 

<img width="503" height="544" alt="Image" src="https://github.com/user-attachments/assets/082299fc-c465-4b94-9412-0d85f8ed9d18" />

**Comment by usert5432** [comment]
Hi @nikky4D,

> Could you elaborate on the finetuning and evaluation process? what is the command to start with a pretrained model for finetuning? Any advice on how best to finetune?

Sure. Let me elaborate on the evaluation first. This is the part of the code that handles each evaluation step: https://github.com/realtime-intelligence/evrt-detr/blob/02811f12705e9a586d18a6c7baf2c9796ed9bb52/evlearn/models/frame_detection_rtdetr.py#L217-L238

You may want to try adding there print/debug statements to double check that the validation dataset contains labels, e.g.:
```
print(self._data.labels)
```

If it does, then either model fails to detect any objects correctly, or they are filtered by the evaluator. You can examine model predictions by looking at the `rtdetr_preds` object (line 230). If it is empty, then perhaps the model is not trained enough/overfitting.

Next, this is the code that calls COCO evaluator: [`evlearn/eval/evaluator/psee_evaluator.py`](https://github.com/realtime-intelligence/evrt-detr/blob/main/evlearn/eval/evaluator/psee_evaluator.py)

In particular, you can try tracing this function: https://github.com/realtime-intelligence/evrt-detr/blob/02811f12705e9a586d18a6c7baf2c9796ed9bb52/evlearn/eval/evaluator/psee_evaluator.py#L32-L36

It receives a list of ground truth and predicted boxes. Then, it applies prohesee filters (line 63-64). If these filters are too aggressive for your case -- you can try simply commenting it out.

Finally, it calls COCO evaluator (line 71). If may also return 0's if the overlap between ground truth and predicted boxes is too small.

**Comment by usert5432** [comment]
To fine-tune a model, you can simply add `transfer` parameter of the configuration (defined by `args_dict`) and set it to:
```
    'transfer'   : {
        'base_model' : TRANSFER_PATH,
        'transfer_map' : {
            'backbone' : 'ema_backbone',
            'encoder'  : 'ema_encoder',
            'decoder'  : 'ema_decoder',
            'ema_backbone' : 'ema_backbone',
            'ema_encoder'  : 'ema_encoder',
            'ema_decoder'  : 'ema_decoder',
        },
        'fuzzy'   : 'ignore-mismatching-shapes',
        'strict'  : True,
        'load_train_state'    : False,
        'use_last_checkpoint' : False,
    },
```

where TRANSFER_PATH is a path to a pre-trained model on the disk. You can download our pre-trained model from here https://github.com/realtime-intelligence/evrt-detr?tab=readme-ov-file#2-optional-downloading-pre-trained-models

**Comment by nikky4D** [comment]
Thank you very much. These are very helpful. I'll start working with them and let you know how it goes. I'll close the comment for now.

**Comment by nikky4D** [comment]
Just an update, I was able to train on my data. Using pretrained model, I was able to start seeing changes in validation after training for at least 30-40 epochs.

**Comment by usert5432** [comment]
Glad it is starting to work! Please let me know if you experience further issues.
